subsys/instrumentation/common/instr_common.c

```diff
@@ -0,0 +1,565 @@
+/*
+ * Copyright 2023 Linaro
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/instrumentation/instrumentation.h>
+#include <zephyr/instrumentation/instrumentation_timestamp.h>
+#include <zephyr/instrumentation/instrumentation_ringbuffer.h>
+#include <zephyr/kernel.h>
+#include <zephyr/kernel_structs.h>
+#include <zephyr/arch/arm/aarch32/cortex_m/cmsis.h>
+
+#include <zephyr/sys/reboot.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+
+#include <zephyr/drivers/retained_mem.h>
+
+#include <kernel_internal.h>
+#include <ksched.h>
+
+/* Memory buffer to store instrumentation event records has the following modes:
+
+ - Callgraph ring buffer (default): Replace oldest entry when buffer
+ full, at the cost of processing stray entry events that have been
+ truncated.
+
+ - Callgraph fixed buffer: Stop buffering events when the buffer is
+ full, ensuring we have a callgraph from reset or from whenever the
+ trigger event was entered.
+
+ - Statistical: Buffer functions until out of memory
+
+*/
+
+const static struct device *retained_mem_dev = DEVICE_DT_GET(DT_ALIAS(retainedmemdevice));
+
+static uint8_t _instr_initialized = 0;
+static uint8_t _instr_on = 0;
+static uint8_t _instr_enabled = 0;
+static uint8_t _instr_tracing_disabled = 0;
+static uint8_t _instr_profiling_disabled = 0;
+
+// For discovered functions. The functions are added to 'disco_func' as they
+// are called in the execution flow, hence "discovered" functions.
+struct fentry_ {
+	void* func_addr;
+
+	int time_stack_pos;
+	uint32_t time_stack[16];
+	uint32_t delta_t;
+};
+
+typedef struct fentry_ fentry;
+
+#define NUM_DISCO_FUNCS 256
+int disco_func_pos = 0;
+fentry disco_func[NUM_DISCO_FUNCS] = { 0 };
+
+// Number of unbalanced entries/exists pair
+int unbalanced = 0;
+
+// To control counter (cycles, timer, etc) initialization
+int cntr_initialized = 0;
+
+#define RETAINED_RAM_MAGIC	0xBEEF
+
+uint16_t magic = 0xABBA;
+
+extern void INSTR_TRIGGER_FUNCTION(void);
+extern void INSTR_STOPPER_FUNCTION(void);
+
+void* k_trigger_callee = INSTR_TRIGGER_FUNCTION;
+void* k_stopper_callee = INSTR_STOPPER_FUNCTION;
+
+void* trigger_callee;
+void* stopper_callee;
+
+/*
+ * TODO(gromero): remove hack and use new RETAINED_MEM_MUTEXES=n config.
```

Comment:
Likewise retention subsystem has a similar Kconfig, note that both retained memory and retention subsystem mutex Kconfigs would need to be set to n

Reply:
@nordicjm Hi. I was not able to find any Kconfig syntax that would allow forcing `RETAINED_MEM_MUTEXES=n`
and `RETENTION_MUTEXES=n` from `subsys/instrumentation/Kconfig`. I put them in `prj,conf` (which works). Any idea on how to set them in `Kconfig`?

Reply:
Indeed it is not possible to de-select a Kconfig from a Kconfig, I will need to rework how it is done

Reply:
Have posted this: https://github.com/zephyrproject-rtos/zephyr/issues/59254 will have a look at it after the 3.4 release is complete

Reply:
@nordicjm Hi Jamie. Thank you!